### DISCUSSION TITLE
Where should I define indexes if I'm defining Model separately?

### DISCUSSION THREAD

**Discussion by rjmackay** [sensedeep/dynamodb-onetable#150] [Q&A] [answered]
If I define models separately, where's the best place to define the indexes?
Does this look correct?
```
const table = new table({
  client: client,
  name: TABLE_NAME,
  schema: {
    indexes: {
      primary: { hash: 'pk', sort: 'sk' },
      // Used for pull claimed by driverId OR unclaimed by zone
      gs1: { hash: 'gs1pk', sort: 'gs1sk', project: 'all' },
    }
    models: {}
  },
})

const Card = new Model(table, 'Card', {
    fields: {
    /* Model schema field definitions */
    }
})
```

**Comment by mobsense** [comment, marked as answer]
Yes, that is fine to define the indexes as part of the schema passed to the table.

However, there is a big benefit of defining your models statically via the schema and not by using the Model constructor. When you use an explicit Model constructor, the model is not added to the schema unless you call addModel.

SenseDeep (and others) are creating a suite of DynamoDB tools that rely on being able to read a schema that statically documents the data schema. So having a schema be the central source of truth regarding your entities and indexes is really beneficial. The schema can then be stored in the table for access by various tools. 

Specifically, if you have a schema defined, you can create intelligent data browsers that understand the entities. You can then create tools that can query by entity, follow inter-entity references, and net/net understand the data stored in your table more deeply.

What we do here is have a single Schema.js that defines the indexes and models. We then use getModel() to retrieve the models as required. In some factory code, we use db.API("ModelName", ...) . This too requires a pre-defined schema.

**Reply by rjmackay** [reply]
Gotcha. I was using a single central schema in one file, along with helper methods, but it's just been getting unwieldy having it all in one file. So I was hoping to split up the schema and co-locate it with the relevant helper methods.

What's the disadvantage of a split schema if I call addModel? Does that still rule out using some of the other tooling?

**Reply by mobsense** [reply]
A split schema cannot be read by a 3rd party tool. Your code and your use will be fine. But migration tools, smart browsers, and single-table schema designers require the ability to read the schema. Doing it separately can work for you, but you then need to create / maintain a combined central schema if you want to use any of these new tools.

I can talk more deeply about what SenseDeep is doing, but I expect that others will also. We'll be releasing in a few weeks a suite of DynamoDB tools that includes: a single-table browser and editor, a schema designer, migration manager and detailed single-table metrics. The ideas is that it helps you create your single-table designs and then helps you manage and evolve them. Instead of peering at raw DynamoDB data with encoded keys and packed attributes, these tools understand the schema and can present things as entities with relationships between entities. We've used these tools in house over the past year and it greatly helps evolve single-table designs.

**Reply by rjmackay** [reply]
On a related note, is there any way to set per-model options without defining models separately? ie. I want to enable `timestamps` for some models but not others

**Reply by mobsense** [reply]
Not yet, but that is a good idea.  I would imagine that timestamps, nulls, createdField, updatedField would be possible fields to override. Need to think of a good way to specify that.
